### ISSUE TITLE
orphaned extension process prevents proper restart/shutdown

### ISSUE THREAD

**Issue by groob** [kolide/launcher#341]
I ran into an issue yesterday, where an autoupdate to the osqueryd process has updated osqueryd, but livequery was still showing the old version hanging until I rebooted my machine. After digging around a bit I can reproduce.

First, look at the list of running processes:

```
ps axo user,pid,ppid,args |awk '/launcher|osquery/'
victor           22949  5714 ./build/launcher --args
victor           23326 22949 /usr/local/bin/osqueryd --args
victor           23327 23326 /Users/victor/go/src/github.com/kolide/launcher/build/osquery-extension.ext --args
```

As expected there are 3 processes; launcher, osqueryd and the exension socket. 
Now I run `kill -9 23326`, which kills the osqueryd process. If everything goes well, osqueryd should restart. Instead I see the following output. 

```
victor           22949  5714 ./build/launcher --args
victor           23327     1 /Users/victor/go/src/github.com/kolide/launcher/build/osquery-extension.ext --args
```

Instead of terminating the extension, which is owned by osqueryd, that process is ghosted (ppid is now 1). 

After separately killing `23327` osqueryd starts up healthy again.

**Comment by groob** [comment]
Note: If you kill osqueryd and wait long enough eventually the ghost process goes away and osqueryd does start up.

**Comment by groob** [comment]
Killing launcher instead of osqueryd also leaves a ghost extension socket running. Doing that multiple times leaves multiple running extension processes.

**Comment by groob** [comment]
I did some additional testing and found out that the runtime kills the entire process group when osqueryd is killed. 

This appears to work reliably:
https://github.com/kolide/launcher/blob/ee6f6f774c94a0fed95f92b0dbc38f65d4689734/pkg/osquery/runtime/runtime.go#L469-L499

**Comment by groob** [comment]
Looking more into this, it's actually a regression from 
https://github.com/kolide/launcher/commit/74e73f5f0474c47aaee0f18b9071cc5099c7e6cb

The restart and shutdown methods are now a noop because the 'Runner" always appears to be nil:
https://github.com/kolide/launcher/blob/ee6f6f774c94a0fed95f92b0dbc38f65d4689734/cmd/launcher/extension.go#L146-L162

**Comment by groob** [comment]
I fixed the initial issue I experience and debugged in #343. 

The fact remains that if you `kill -9` the launcher process itself, the extension socket remains orphaned. For that I'm leaving #341 as a possible solution.

**Comment by directionless** [comment]
I think this was likely fixed by #342. Closing for now
